File: packages/sdk-js/src/utils/pagination.ts
import { PaginatedResponse, ListParams } from '../types/common.types';

/**
 * Async iterator for paginating through all results
 */
export async function* paginate<T, P extends ListParams>(
  fetchPage: (params: P) => Promise<PaginatedResponse<T>>,
  params: P
): AsyncGenerator<T, void, undefined> {
  let cursor: string | undefined;
  let hasMore = true;

  while (hasMore) {
    const response = await fetchPage({
      ...params,
      cursor,
    });

    for (const item of response.data) {
      yield item;
    }

    hasMore = response.pagination.hasMore;
    cursor = response.pagination.cursor;
  }
}

/**
 * Collect all items from a paginated endpoint
 */
export async function collectAll<T, P extends ListParams>(
  fetchPage: (params: P) => Promise<PaginatedResponse<T>>,
  params: P,
  maxItems?: number
): Promise<T[]> {
  const items: T[] = [];

  for await (const item of paginate(fetchPage, params)) {
    items.push(item);
    if (maxItems && items.length >= maxItems) {
      break;
    }
  }

  return items;
}

/**
 * Create a paginated list helper
 */
export function createPaginatedList<T, P extends ListParams>(
  fetchPage: (params: P) => Promise<PaginatedResponse<T>>
) {
  return {
    /**
     * Fetch a single page of results
     */
    list: (params: P) => fetchPage(params),

    /**
     * Iterate through all results
     */
    listAll: (params: Omit<P, 'cursor'>) =>
      paginate(fetchPage, params as P),

    /**
     * Collect all results into an array
     */
    collectAll: (params: Omit<P, 'cursor'>, maxItems?: number) =>
      collectAll(fetchPage, params as P, maxItems),
  };
}

File: packages/sdk-python/constitutional/utils/pagination.py
"""Pagination utilities for the Constitutional SDK."""

from typing import TypeVar, Generator, Callable, Optional
from constitutional.types.common import PaginatedResponse


T = TypeVar("T")


def paginate(
    fetch_page: Callable[..., PaginatedResponse[T]],
    **kwargs: object,
) -> Generator[T, None, None]:
    """
    Iterate through all pages of a paginated endpoint.

    Args:
        fetch_page: Function to fetch a page of results
        **kwargs: Arguments to pass to fetch_page

    Yields:
        Individual items from each page
    """
    cursor: Optional[str] = None
    has_more = True

    while has_more:
        response = fetch_page(cursor=cursor, **kwargs)

        for item in response.data:
            yield item

        has_more = response.pagination.has_more
        cursor = response.pagination.cursor


def collect_all(
    fetch_page: Callable[..., PaginatedResponse[T]],
    max_items: Optional[int] = None,
    **kwargs: object,
) -> list[T]:
    """
    Collect all items from a paginated endpoint into a list.

    Args:
        fetch_page: Function to fetch a page of results
        max_items: Maximum number of items to collect (optional)
        **kwargs: Arguments to pass to fetch_page

    Returns:
        List of all items
    """
    items: list[T] = []

    for item in paginate(fetch_page, **kwargs):
        items.append(item)
        if max_items and len(items) >= max_items:
            break

    return items
